docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,76 @@
+# Generics parameters of ByRefLike types
+
+Using ByRefLike types in Generic parameters is possible by building upon support added for `ref` fields. Scenarios that would benefit most from this are those involving `Span<T>`. For example, consider the following examples:
+
+- `Span<TypedReference>` &ndash; Represents the general case where a ByRefLike type is used as a Generic parameter. This specific case would be desirable for a more efficient Reflection API.
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; This IL sequence must resolve to a method implemented on `object`, or a default interface method.
+
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`.
```

Comment:
Types like `Span<T>` have several APIs that conflict with the rules above. Most notable:

```c#
public Span(T[]? array);
public static implicit operator Span<T>(T[]? array);
```

How are we going to rationalize these APIs when we add `allow T: ref struct` to `Span<T>`?

Reply:
Jared and I have discussed this separately before. There are probably multiple viable ways to go but I think that exposing a compiler-known way to suppress constraint checks on specific APIs would do it. Then, I'm hoping that conversion errors at the use site will prevent the APIs from being used when the type arguments don't permit it.

```cs
using System;

namespace System.Diagnostics.CodeAnalysis
{
    public sealed class SuppressConstraintChecksAttribute : Attribute { }
}

ref struct Span<T> allow T : ref struct
{
    [SuppressConstraintChecks]
    public Span(T[]? array) { ... }

    [SuppressConstraintChecks]
    public static implicit operator Span<T>(T[]? array) { ... }
}

class Program
{
    static void M(Span<Span<int>> span)
    {
        var span1 = new Span<Span<int>>(null); // error: can't convert 'null' to type 'Span<int>[]'
        var span2 = (Span<Span<int>>)null; // error: can't convert 'null' to type 'Span<int>[]'
    }
}
```

This might leave a hole where the attribute could be used on methods which only violate type parameter constraints within the implementation, not in the signature. The user wouldn't be stopped from doing it at compile time but could get a runtime error. That might just be a case of "play stupid games, win stupid prizes".

Tagging @davidwrighton who had interest in a 'SuppressConstraintChecksAttribute' for a scenario with static interface members, IIRC.

Reply:
> That might just be a case of "play stupid games, win stupid prizes".

That is how I am thinking of this feature. There are cases where the compiler can definitely determine if the user is doing something wrong, but also where it _might_ be doing something wrong. The [`Span<T>.ToArray()`](https://docs.microsoft.com/dotnet/api/system.span-1.toarray) is another example. If a user calls `ToArray()` and `T` is known to be ByRefLike, the compiler could issue a failure. However, if `T` isn't know at compile time then it would be up to the implementer to have handled this appropriately using something like the proposed intrinsi-fication of `typeof(T.).IsByRefLike` in this document. The updated `Span<T>.ToArray()` would look as follows and handle the case at runtime. The other APIs in question would have similar checks.

```diff
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public T[] ToArray()
{
+    if (typeof(T).IsByRefLike)
+        throw new InvalidProgramException();

    if (_length == 0)
        return Array.Empty<T>();

    var destination = new T[_length];
    Buffer.Memmove(ref MemoryMarshal.GetArrayDataReference(destination), ref _pointer.Value, (uint)_length);
    return destination;
}
```

Reply:
It will be enforced but can all be detected at JIT time, which can then error once `T` is defined. The fact that `typeof(T).IsByRefLike` is a JIT intrinsic will permit it to detect that branch is always taken and the remaining instructions can be removed. In the event the user didn't put in that call the invalid instructions would then be processed but during the inspection of `T` it would be determined something is invalid and `InvalidProgramException` throw.

I suppose the example is bad since it is also throwing. Let me update it to reflect a more reasonable return.

Reply:
Actually `ToArray` is a hard example to make something more reasonable. Imagine the return could be something that is used by the caller in a reasonable way. Unsure what that would be in this case though.

Reply:
Thanks @RikkiGibson and @jaredpar. I'd not fully appreciated the compiler issues here.